runtime-community: Failure to build System.Security.Cryptography.Native/opensslshim.c due to overflow in expression with type long

## Build Information

Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=728088
Build error leg or test failing: `runtime-community` - `linux-armv6 Release AllSubsets_Mono`

## Error Message

```json
{
  "ErrorMessage" : [
    "overflow in expression; result is -2147483648 with type 'long' [-Werror,-Winteger-overflow]",
    "time_t timeVal = (time_t)INT_MAX + 1"
  ],
  "BuildRetry" : false,
  "ExcludeConsoleLog" : false
}
```

- PR: Official rolling build of runtime-community in `main`: https://dev.azure.com/dnceng-public/public/_build/results?buildId=728088&view=results
- Queue: `linux-armv6 Release AllSubsets_Mono`
- Job result: https://dev.azure.com/dnceng-public/public/_build/results?buildId=728088&view=logs&j=db8e93a1-262e-5037-1aaf-a14621311ed2&t=df0c0844-666d-543b-d576-d26fca1aab72&l=3504
- Output:
```
[ 88%] Building C object System.IO.Compression.Native/CMakeFiles/System.IO.Compression.Native.dir/__w/1/s/src/native/external/brotli/enc/bit_cost.c.o
/__w/1/s/src/native/libs/System.Security.Cryptography.Native/opensslshim.c(238,38): error G7D1C1CEA: overflow in expression; result is -2147483648 with type 'long' [-Werror,-Winteger-overflow] [/__w/1/s/src/native/libs/build-native.proj]
##[error]src/native/libs/System.Security.Cryptography.Native/opensslshim.c(238,38): error G7D1C1CEA: (NETCORE_ENGINEERING_TELEMETRY=Build) overflow in expression; result is -2147483648 with type 'long' [-Werror,-Winteger-overflow]
      time_t timeVal = (time_t)INT_MAX + 1;
                                       ^
  1 error generated
```
<!-- Known issue validation start -->
 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=728088
**Error message validated:** `[overflow in expression; result is -2147483648 with type 'long' [-Werror,-Winteger-overflow] time_t timeVal = (time_t)INT_MAX + 1`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 7/2/2024 10:30:44 PM UTC
<!-- Known issue validation end -->
<!--Known issue error report start -->

### Report
|Build|Definition|Step Name|Console log|Pull Request|
|---|---|---|---|---|
|[728642](https://dev.azure.com/dnceng-public/public/_build/results?buildId=728642)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/728642/logs/41)|dotnet/runtime#102403|
|[728088](https://dev.azure.com/dnceng-public/public/_build/results?buildId=728088)|dotnet/runtime|Build product|[Log](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/728088/logs/23)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|2|
<!--Known issue error report end -->

--- a/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
+++ b/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
@@ -233,9 +233,9 @@ void InitializeOpenSSLShim(void)
     }
 
 #if defined(TARGET_ARM) && defined(TARGET_LINUX)
-    // This value will represent a time in year 2038 if 64-bit time is used,
-    // or 1901 if the lower 32 bits are interpreted as a 32-bit time_t value.
-    time_t timeVal = (time_t)((unsigned long)INT_MAX + 1u);
+    // This value will represent a time in year 2038 if 64-bit time_t is used,
+    // or 1901 if the conversion overflowed.
+    time_t timeVal = (time_t)0x80000000U;
     struct tm tmVal = { 0 };
 
     // Detect whether openssl is using 32-bit or 64-bit time_t.

--- a/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
+++ b/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
@@ -235,7 +235,7 @@ void InitializeOpenSSLShim(void)
 #if defined(TARGET_ARM) && defined(TARGET_LINUX)
     // This value will represent a time in year 2038 if 64-bit time is used,
     // or 1901 if the lower 32 bits are interpreted as a 32-bit time_t value.
-    time_t timeVal = (time_t)INT_MAX + 1u;
+    time_t timeVal = (time_t)((unsigned long)INT_MAX + 1u);
     struct tm tmVal = { 0 };
 
     // Detect whether openssl is using 32-bit or 64-bit time_t.

--- a/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
+++ b/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
@@ -235,7 +235,7 @@ void InitializeOpenSSLShim(void)
 #if defined(TARGET_ARM) && defined(TARGET_LINUX)
     // This value will represent a time in year 2038 if 64-bit time is used,
     // or 1901 if the lower 32 bits are interpreted as a 32-bit time_t value.
-    time_t timeVal = (time_t)INT_MAX + 1;
+    time_t timeVal = (time_t)INT_MAX + 1u;
     struct tm tmVal = { 0 };
 
     // Detect whether openssl is using 32-bit or 64-bit time_t.

--- a/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
+++ b/src/native/libs/System.Security.Cryptography.Native/opensslshim.c
@@ -233,8 +233,10 @@ void InitializeOpenSSLShim(void)
     }
 
 #if defined(TARGET_ARM) && defined(TARGET_LINUX)
-    // This value will represent a time in year 2038 if 64-bit time_t is used,
-    // or 1901 if the conversion overflowed.
+    c_static_assert_msg(sizeof(time_t) == 8, "Build requires 64-bit time_t.");
+    
+    // This value will represent a time in year 2038 if 64-bit time is used,
+    // or 1901 if the lower 32 bits are interpreted as a 32-bit time_t value.
     time_t timeVal = (time_t)0x80000000U;
     struct tm tmVal = { 0 };
 

